### PR TITLE
All DB2 changes

### PR DESCRIPTION
**Pull request by dwadler**

### CODE REVIEW

**Comment by luipir** [comment]
Hi I didn't read all the code... just a flythrough to see that you should use the new centralised auth system with QgsAuthManager. Be inspired by Posggres provider in current master.

**Comment by m-kuhn** [comment]
Hi @dwadler 

Good stuff!

To check why the build fails you can have a look at the [build log here](http://dash.orfeo-toolbox.org/index.php?project=QGIS&filtercount=1&showfilters=1&field1=buildname/string&compare1=63&value1=573d5e9b225efcc2e532d21334dc85e622938096&date=2016-01-29)

I wonder, is there a special SDK required to compile this?
I assume it's not possible to build it on our CI with DB2 support enabled due to lack of the SDK. And it won't be possible to run tests there because of the lack of a DB to test against. I would still recommend to run the basic provider tests for DB2 (just copy https://github.com/qgis/QGIS/blob/master/tests/src/python/test_provider_mssql.py, adjust it, put the reference dataset into your DB, add it to the testsuite and run the testsuite). This will give you some confidence that certain standards are followed.

It should be opt-in (with an appropriate cmake option) in either case.

**Comment by nyalldawson** [comment]
I think you *can* use db2 on Travis, at least going by https://github.com/SOCI/soci/blob/master/bin/ci/before_install_db2.sh and and https://github.com/angoca/db2-travis . (Haven't looked too closely though)

File: src/providers/db2/qgsdb2featureiterator.cpp [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+                    qgsDoubleToString( request.filterRect().xMaximum() ),
+                    qgsDoubleToString( request.filterRect().yMaximum() ),
+                    QString::number( mSource->mSRId ) );
+
+#endif
+    filterAdded = true;
+  }
+
+  // set fid filter
+  if ( request.filterType() == QgsFeatureRequest::FilterFid && !mSource->mFidColName.isEmpty() )
```
Comment (nyalldawson) [inline src/providers/db2/qgsdb2featureiterator.cpp]: I'd *strongly* suggest adding handling of FilterFids requests too.. this is a huge performance win across lots of different operations

**Comment by dwadler** [comment]
Wow - I'm amazed at all the quick responses. I didn't realize it was going to get such exposure. I was just hoping to do a "dry run" of the process.

It should be possible to build without any DB2 SDK - we are just using the Qt database support. It won't actually run without a live DB2 connection.

I am in the process of testing with python/test_provider_db2.py - still need to sort out a couple tests that are failing.

Thank you for the link to the build output - I wasn't sure how to find it.

Is this one of the errors that caused it to fail?
/.../QGIS/src/providers/db2/qgsdb2provider.h:94: Note: No relevant classes found. No output generated.
I don't understand as it starts with:
class QgsDb2Provider : public QgsVectorDataProvider
The code runs fine in my environment despite this.

If there were no errors, would this have automatically committed into QGIS master?
I'm not ready for that.

File: src/providers/db2/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,43 @@
+
+########################################################
+# Files
+
+SET(DB2_SRCS
+  qgsdb2provider.cpp qgsdb2dataitems.cpp qgsdb2newconnection.cpp qgsdb2tablemodel.cpp qgsdb2featureiterator.cpp qgsdb2sourceselect.cpp
+)
+SET(DB2_MOC_HDRS
+  qgsdb2provider.h qgsdb2dataitems.h qgsdb2newconnection.h qgsdb2tablemodel.h qgsdb2featureiterator.h qgsdb2sourceselect.h
```
Comment (nyalldawson) [inline src/providers/db2/CMakeLists.txt]: qgsdb2provider.h and qgsdb2featureiterator.h shouldn't be listed here - only headers with the Q_OBJECT macro need to be included in this section. That's what's causing the build errors.

**Comment by nyalldawson** [comment]
You'll also need to clear all the warnings from http://dash.orfeo-toolbox.org/viewBuildError.php?type=1&buildid=216071 and http://dash.orfeo-toolbox.org/viewBuildError.php?type=1&buildid=216065 (we have a no-warnings-allowed policy to keep the codebase clean).

> If there were no errors, would this have automatically committed into QGIS master?
I'm not ready for that.

Definitely not! It still needs to be reviewed, accepted and merged by someone with commit rights.

**Comment by nyalldawson** [comment]
Just saw the strange clang build failures your experiencing on Travis - I suggest rebasing this branch against master, as those failures were fixed in a commit ~ 2/3 weeks ago.

Hope that helps!

**Comment by dwadler** [comment]
Where should I be looking for the errors?  I don't see anything at
https://api.travis-ci.org/jobs/110831331/log.txt?deansi=true

I thought I merged QGIS master into our branch yesterday and just did it again a few minutes ago.

**Comment by nyalldawson** [comment]
I was looking here: https://travis-ci.org/qgis/QGIS/jobs/110796253 and the resultant build log http://dash.orfeo-toolbox.org/viewBuildError.php?buildid=218689

**Comment by dwadler** [comment]
I don't understand the 3 errors that we are getting - there doesn't seem to be anything related to the DB2 code, at least that I can see.  All the warnings should be fixed now.

**Comment by m-kuhn** [comment]
There are formatting problems in the code (e.g. whitespaces at the end of line)

Just run the following script with the files for which you get indentation problems:

    ./scripts/astyle.sh src/app/qgisapp.cpp

https://travis-ci.org/qgis/QGIS/jobs/110959435#L1135

And then there's a new member which requires python (sip) bindings. https://travis-ci.org/qgis/QGIS/jobs/110959435#L2089

You need to add this method to python/core/qgsdatasourceuri.sip.

**Comment by m-kuhn** [comment]
Have a look here, the main issues are missing `override`s.

http://dash.orfeo-toolbox.org/viewBuildError.php?type=1&buildid=220670

    virtual QgsFeatureIterator getFeatures( const QgsFeatureRequest& request = QgsFeatureRequest() );

should be

    virtual QgsFeatureIterator getFeatures( const QgsFeatureRequest& request = QgsFeatureRequest() ) override;

and so on

The other build errors are unrelated to your changes and no reason for you to worry.

To get db2 tests in CI (travis) but it will require that somebody digs into the details of what's required to spin a db2 server up and compile against the db2 sdk. It would of course be great if we could run them there but I wouldn't say it's mission critical.

**Comment by dwadler** [comment]
The override warnings are strange - I could have sworn that they didn't appear before.  We will take care of these.

It isn't necessary to compile against the DB2 SDK as we are using the ODBC driver, so DB2 support can be included in the build without any problem.

Unfortunately, it isn't as straightforward to test as DB2 needs to be installed and customized to include the spatial support. (not technically difficult but may not be simple procedurally)

**Comment by m-kuhn** [comment]
You probably caught a bad timing for the last test, we've had some troubles in master.

This is the interesting part for you:

```
/.../QGIS/src/providers/db2/qgsdb2expressioncompiler.cpp:107:14: warning: enumeration value 'uoMinus' not handled in switch [-Wswitch]
     switch ( n->op() )
             ^
```

http://dash.orfeo-toolbox.org/viewBuildError.php?type=1&buildid=221084

**Comment by dwadler** [comment]
Closing this PR as branch db2-development seems hopelessly messed up in git.
Opening a new PR against db2-development215 which has all DB2 provider commits merged.

**Pull request closed without merging**